rasterio/_warp.pyx

```diff
@@ -653,8 +653,9 @@ cdef class WarpedVRTReaderBase(DatasetReaderBase):
         self.dst_crs = dst_crs
         self.resampling = resampling
         self.tolerance = tolerance
-        self.src_nodata = src_nodata
-        self.dst_nodata = dst_nodata
+
+        self.src_nodata = (self.src_dataset.nodata or 0.0) if src_nodata is None else src_nodata
+        self.dst_nodata = (self.src_nodata or 0.0) if dst_nodata is None else dst_nodata
```

Comment:
@sgillies Whats the reasoning behind defaulting to `0` instead of `None`?

Reply:
0 is the GDAL default for destination nodata, but you're right, it's not so for the source. I'll see about switching to None.